massa-bootstrap/src/server.rs

```diff
@@ -14,363 +42,451 @@ use massa_signature::KeyPair;
+    config: BootstrapConfig,
+    mut establisher: Establisher,
     keypair: KeyPair,
     version: Version,
-) -> Result<Option<BootstrapManager>, BootstrapError> {
+) -> Result<Option<BootstrapManager>, Box<BootstrapError>> {
     massa_trace!("bootstrap.lib.start_bootstrap_server", {});
-    if let Some(bind) = bootstrap_config.bind {
-        let (manager_tx, manager_rx) = mpsc::channel::<()>(1);
+    let Some(listen_addr) = config.listen_addr else {
+        return Ok(None);
+    };
 
-        let join_handle = tokio::spawn(async move {
-            BootstrapServer {
-                consensus_controller,
-                network_command_sender,
-                final_state,
-                establisher,
-                manager_rx,
-                bind,
-                keypair,
-                version,
-                ip_hist_map: HashMap::with_capacity(bootstrap_config.ip_list_max_size),
-                bootstrap_config,
-            }
-            .run()
-            .await
-        });
-        Ok(Some(BootstrapManager {
-            join_handle,
-            manager_tx,
-        }))
-    } else {
-        Ok(None)
-    }
+    // TODO(low prio): See if a zero capacity channel model can work
+    let (stopper_tx, stopper_rx) = crossbeam::channel::bounded::<()>(1);
+
+    let runtime = Runtime::new().expect("Failed to create a bootstrap runtime");
+    let listener = establisher
+        .get_listener(listen_addr)
+        .await
+        .map_err(BootstrapError::IoError)?;
+
+    let Ok(max_bootstraps) = config.max_simultaneous_bootstraps.try_into() else {
+        return Err(BootstrapError::GeneralError("Fail to convert u32 to usize".to_string()).into());
+    };
+    // This is the primary interface between the async-listener, and the (soon to be) sync worker
+    let (listener_tx, listener_rx) =
+        crossbeam::channel::bounded::<(Duplex, SocketAddr)>(max_bootstraps);
+
+    let allow_block_list = SharedAllowBlockList::new(
+        config.bootstrap_whitelist_path.clone(),
+        config.bootstrap_blacklist_path.clone(),
+    )
+    .map_err(BootstrapError::GeneralError)?;
+
+    let update_handle = runtime.handle().clone().spawn(BootstrapServer::run_updater(
+        allow_block_list.clone(),
+        config.cache_duration.into(),
+    ));
+    let listen_handle = runtime
+        .handle()
+        .clone()
+        .spawn(BootstrapServer::run_listener(listener, listener_tx));
+
+    let main_handle = std::thread::spawn(move || {
+        BootstrapServer {
+            consensus_controller,
+            network_command_sender,
+            final_state,
+            listener_rx,
+            stopper_rx,
+            allow_block_list,
+            keypair,
+            version,
+            ip_hist_map: HashMap::with_capacity(config.ip_list_max_size),
+            bootstrap_config: config,
+            runtime,
+        }
+        .run_loop(max_bootstraps)
+    });
+    // does `runtime` get dropped here?
```

Comment:
Not so sure about this one. Perhaps it's lifetime is bound to the join-handles that were spawned from this runtime?

Reply:
You moved it to the thread of the bootstrap server no ?

Reply:
Yeah.

One pain-point was that it needs to drop in a non-blocking context

a) using the runtime to send errors back to the client from a spawned thread frees up some nasty blocks
b) the server drop coincides with a good time to drop the runtime.